### PR TITLE
(PUP-7879) configure runtime for tasks

### PR DESCRIPTION
**Pull request by hlindberg**
This is the first PR in an expected series of PRs for PUP-7879.

In this PR a "puppet script" application is added that acts like "puppet apply" but it does not produce a catalog. This application uses a ScriptCompiler to do the compilation.
Other than just skipping doing lots of things that are only relevant when compiling a catalog, this configuration simplifies some paths through the logic.
There are a few commits with small changes to avoid unwanted side effects performed by the regular logic. Some of those can probably be refactored by refactoring the original code (logic in the wrong place typically).

One change is in Scope.to_s as it generates "Scope(#{@resource})" when being logged. As that is irrelevant (there are no classes or other resource) a change was made there to instead pick the location from the puppet stack - that path is then reduced somewhat in length for paths pointing into the environment in use. More can be done with that to also shorten paths into modules and show <module>name/...

It is likely that there are additional opportunities to clean up code in the ScriptCompiler. All of the exceptions in the methods are there to ensure that general purpose evaluation does not by accident end up calling methods for unsupported features. If we truly know they are never called we can just remove all of those. The work is also a precursor to refactoring compiler into a base class that is some kind of "Executor" (i.e. Evaluator + Context (loaders etc).

### CODE REVIEW

File: lib/puppet/parser/script_compiler.rb [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+  def add_catalog_validators
+    raise _('use of catalog validators is not supported when scripting')
+  end
+
+  # Return a list of all of the defined classes.
+  def_delegator :@catalog, :classes, :classlist
+
+  def with_context_overrides(description = '', &block)
+    Puppet.override( @context_overrides , description, &block)
+  end
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: This method looks exactly the same in the super class.

File: lib/puppet/parser/script_compiler.rb [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+  def add_catalog_validator(catalog_validators)
+    raise _('use of catalog validators is not supported when scripting')
+  end
+
+  def add_catalog_validators
+    raise _('use of catalog validators is not supported when scripting')
+  end
+
+  # Return a list of all of the defined classes.
+  def_delegator :@catalog, :classes, :classlist
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: Declared in super too.

File: lib/puppet/parser/script_compiler.rb [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+      Puppet::Util::Profiler.profile(_("Script: Created settings scope"), [:compiler, :create_settings_scope]) { create_settings_scope }
+
+      #TRANSLATORS "main" is a function name and should not be translated
+      Puppet::Util::Profiler.profile(_("Script: Evaluated main"), [:compiler, :evaluate_main]) { evaluate_main }
+    end
+  end
+
+  def validate_catalog(validation_stage)
+    @catalog_validators.select { |vclass| vclass.validation_stage?(validation_stage) }.each { |vclass| vclass.new(@catalog).validate }
+  end
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: Exactly the same in super class.

File: lib/puppet/parser/script_compiler.rb [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+      :loaders  => @loaders,                  # 4x loaders
+    }
+  end
+
+  def_delegator :@collections, :delete, :delete_collection
+
+  # Return the node's environment.
+  def environment
+    node.environment
+  end
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: Exactly the same in super class.

File: lib/puppet/parser/script_compiler.rb [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+  # @api private
+  #
+  def evaluate_site
+    raise _('evaluation of site is not supported when scripting')
+  end
+
+  # @api private
+  def on_empty_site
+    # do nothing
+  end
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: Exactly the same in super class.

File: lib/puppet/parser/script_compiler.rb [hunk 6]
```diff
@@ -0,0 +1,469 @@
...
+
+  # Create a new scope, with either a specified parent scope or
+  # using the top scope.
+  def newscope(parent, options = {})
+    parent ||= topscope
+    scope = Puppet::Parser::Scope.new(self, options)
+    scope.parent = parent
+    scope
+  end
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: Exactly the same in super class.

File: lib/puppet/parser/script_compiler.rb [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+      trusted = Puppet.lookup(:trusted_information) do
+        trusted_param || Puppet::Context::TrustedInformation.local(node)
+      end
+
+      # Ruby 1.9.3 can't apply to_h to a hash, so check first
+      node.trusted_data = trusted.is_a?(Hash) ? trusted : trusted.to_h
+    end
+
+    node
+  end
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: Exactly the same in super class.

File: lib/puppet/parser/script_compiler.rb [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+    # When scripting the trusted data are always local, but set them anyway
+    @topscope.set_trusted(node.trusted_data)
+
+    # Server facts are always about the local node's version etc.
+    @topscope.set_server_facts(node.server_facts)
+
+    # Set $facts for the node running the script
+    facts_hash = node.facts.nil? ? {} : node.facts.values
+    @topscope.set_facts(facts_hash)
+  end
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: Exactly the same in super class.

File: lib/puppet/parser/script_compiler.rb [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+
+  SETTINGS = 'settings'.freeze
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: Defined in super class

File: lib/puppet/parser/script_compiler.rb [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+    #
+    @topscope.merge_settings(environment.name, false)
+  end
+
+  # Return an array of all of the unevaluated resources.  These will be definitions,
+  # which need to get evaluated into native resources.
+  def unevaluated_resources
+    # The order of these is significant for speed due to short-circuiting
+    resources.reject { |resource| resource.evaluated? or resource.virtual? or resource.builtin_type? }
+  end
```
Comment (thallgren) [inline lib/puppet/parser/script_compiler.rb]: Exactly the same in super class.

**Comment by hlindberg** [comment]
The duplication is intentional and there will be a series of changes after this. The "ScriptCompiler" isn't a compiler at all in the same sense as the original and ideally the script compiler should not be derived from Compiler. In fact that may be one of the changes that should be made to decouple the two. This PR as it stands is only the first step. An important factor here is to change as little existing code as possible to make the risks of merging this into puppet as small as possible. Any refactoring of standard puppet should only take place after merge of the feature branch. (There are small changes to standard puppet here, but they should be kept to a minimum).

**Comment by thallgren** [comment]
I understand the importance of not changing existing code. My concern is that once all of this duplicated code has made it into the repository, how do we keep track of it? When is it removed? It's very easy (and also mostly correct) to not fix things that aren't broken and copy/paste in general doesn't break things. So with this approach it's likely that we will end up with lots of copy/pasted stuff that eventually will become a maintenance nightmare. Not sure why that's a better alternative than not changing existing code. It's rarely more expensive to do things right from the start.

File: lib/puppet/application/script.rb [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+    # we do not want the last report to be persisted
+    Puppet::Transaction::Report.indirection.cache_class = nil
+
+    set_log_level
+
+    if Puppet[:profile]
+      @profiler = Puppet::Util::Profiler.add_profiler(Puppet::Util::Profiler::Aggregate.new(Puppet.method(:info), "script"))
+    end
+  end
+end
```
[the remark this answers is not in the document]
Comment (hlindberg) [reply lib/puppet/application/script.rb]: We don't know if we want to keep the Script application

**Comment by hlindberg** [comment]
Pushed more logic that separates ScriptCompiler from Compiler and introduces an AbstractCompiler (mostly to enable Scope to check that it got something it can use even if it is somewhat of a lie). Scope has *many* methods that deal with resources. We cannot refactor that without major breakage, but at least it requires someone to deliberately creating something using AbstractCompiler...

I then checked each and every of the methods and construct in the ScriptCompiler to ensure that it was not used for code evaluation that does not involve resources and classes.
The handling of node / sanitization is there because a refactor of that needs more thought.

**Comment by hlindberg** [comment]
And even more cleanup and changes pushed on. This is starting to look ok now.

File: lib/puppet/parser/abstract_compiler.rb [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+  def qualified_variables
+    raise Puppet::DevError("Class '#{self.class}' should have implemented 'qualified_variables'")
+  end
+
+  # Returns the top scope instance
+  def topscope
+    raise Puppet::DevError("Class '#{self.class}' should have implemented 'topscope'")
+  end
+
+end
```
Comment (thallgren) [inline lib/puppet/parser/abstract_compiler.rb]: Missing newline

**Pull request merged**